### PR TITLE
Create Middleware to Verify Authentication

### PR DESCRIPTION
**Pull request by qwwqwwq**
- Middleware to check if user is authed, eventually we should use this to replace many one-off places in the code were we try to establish authorization
- Used this middleware within the fileClaim controller
- Updated & simplified fileClaimController tests

### CODE REVIEW

File: winterfell/controllers/fileClaimController.js [hunk 1]
```diff
@@ -29,7 +30,7 @@ module.exports = function (app) {
   });
 
   // Get a particular claim
-  app.get('/claims/:extClaimId', function (req, res) {
+  app.get('/claim/:extClaimId', auth.authenticatedOr404, function (req, res) {
```
Comment (ahsu1230) [inline winterfell/controllers/fileClaimController.js]: should be `/claims/...`

Comment (ahsu1230) [inline winterfell/controllers/fileClaimController.js]: well what do you think? We can keep it singular `/claim/:claimId`, but we should be consistent across the board.

File: winterfell/controllers/fileClaimController.js [hunk 2]
```diff
@@ -40,71 +41,68 @@ module.exports = function (app) {
     });
   });
 
-  app.post('/claims/create', function (req, res) {
+  app.post('/claims/create', auth.authenticatedOr404, function (req, res) {
     console.log('[createClaim] request received for ' + JSON.stringify(req.body));
-    var extUserId = req.body.extUserId;
     var callback = function (err, claim) {
       if (claim) {
-        res.status(http.OK).send({claim: FileClaim});
```
Comment (ahsu1230) [inline winterfell/controllers/fileClaimController.js]: my bad that was dumb

File: winterfell/controllers/fileClaimController.js [hunk 3]
```diff
@@ -40,71 +41,68 @@ module.exports = function (app) {
...
+        res.status(http.CREATED).send(claim);
```
Comment (ahsu1230) [inline winterfell/controllers/fileClaimController.js]: in general, I like to always send an object as a response.
Consider when one day, we want to add more sibling objects to claim (let's call it `properties`)
So then we want to send()
```
{
  claim: claim,
  properties: props
}
```

We are more prone to backwards compatibility issues if we are coming from
```
.send(claim)
```
since the front-end will have to change. It gets even more hectic if multiple places on the front-end have to change for this. But then again, this is just a website (deploys instantly unlike apps and extensions) and we probably will only have one place to call this endpoint. Just my two cents from past experiences.

TLDR: Can we keep it as `.send({claim: claim})` please? 👻

File: winterfell/controllers/fileClaimController.js [hunk 4]
```diff
@@ -40,71 +41,68 @@ module.exports = function (app) {
...
       }
-    } else {
-      res.status(http.BAD_REQUEST).send({error: httpErrors.INVALID_CLAIM_ID});
     }
+  }
+
+  app.post('/claim/:extClaimId/submit', function(req, res) {
+    console.log('[submitClaim] request received ' + JSON.stringify(req.body));
+
+    handleClaimStateChange(req.params.extClaimId,
```
Comment (ahsu1230) [inline winterfell/controllers/fileClaimController.js]: nice tyty

File: winterfell/middlewares/auth.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  console.log("Not authed");
+  res.sendStatus(http.NOT_FOUND);
+};
+
+module.exports.authenticatedOrRedirect = function (req, res, next) {
+  if (req.session.userId)
+    return next();
+
+  res.redirect('/');
+};
```
Comment (ahsu1230) [inline winterfell/middlewares/auth.js]: nit: extra line at end of file please
Sorry I'm such a 👿 
http://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

File: winterfell/test/testFileClaimController.js [hunk 6]
```diff
@@ -42,37 +42,31 @@ describe('FileClaimController', function() {
...
+      .get('/claims')
+      .expect(http.NOT_FOUND, done);
   });
 
-  it('Create claim for user - user dne', function(done) {
-    testSession
-      .post('/claims/create')
-      .send({ extUserId: 'asdf' })
-      .expect(http.NOT_FOUND, done);
+  it('should sign in', function (done) {
```
Comment (ahsu1230) [inline winterfell/test/testFileClaimController.js]: this test should not be in fileclaimcontroller test

File: winterfell/middlewares/auth.js [hunk 7]
```diff
@@ -0,0 +1,23 @@
+var http = require('http-status-codes');
+
+function Auth(app) {
+  this.app = app;
+}
+
+module.exports = Auth;
+
+module.exports.authenticatedOr404 = function (req, res, next) {
```
Comment (ahsu1230) [inline winterfell/middlewares/auth.js]: this is awesome

File: winterfell/controllers/fileClaimController.js [hunk 8]
```diff
@@ -40,71 +41,68 @@ module.exports = function (app) {
...
       } else {
-        res.sendStatus(statusCode);
+        res.sendStatus(http.NOT_FOUND);
       }
-    } else {
-      res.status(http.BAD_REQUEST).send({error: httpErrors.INVALID_CLAIM_ID});
     }
+  }
+
+  app.post('/claim/:extClaimId/submit', function(req, res) {
```
Comment (ahsu1230) [inline winterfell/controllers/fileClaimController.js]: `/claim` or `/claims`?

File: winterfell/test/testFileClaimController.js [hunk 9]
```diff
@@ -42,37 +42,31 @@ describe('FileClaimController', function() {
...
+    testSession.post('/auth/login')
+      .send({email: targetUser.email, password: targetUser.password})
+      .expect(200, done);
   });
 
-  it('Create claim for user - success', function(done) {
+
+  it('Create claim for user - user dne', function(done) {
```
Comment (ahsu1230) [inline winterfell/test/testFileClaimController.js]: misleading test name
should be -success case

**Comment by ahsu1230** [comment]
couple of comments, but mostly good.
Let's make sure to decide on `claim/` vs. `claims/` so the front end can change appropriately as well.

File: winterfell/controllers/fileClaimController.js (hunk 1 above)
Comment (qwwqwwq) [inline winterfell/controllers/fileClaimController.js]: the RESTful convention I was aiming for was: use plural for resources that return an array of that model, and singular for resources that return a single instance of the model, so since this is returning just one claim I wanted to change to singular

File: winterfell/controllers/fileClaimController.js (hunk 3 above)
Comment (qwwqwwq) [inline winterfell/controllers/fileClaimController.js]: Ok to the convention is we return an object which is of the form:

`{ <model_name>: <model_object> }`

that sounds reasonable to me

File: winterfell/middlewares/auth.js (hunk 5 above)
Comment (qwwqwwq) [inline winterfell/middlewares/auth.js]: thanks, i agree, sorry I just need to setup my IDE to do this

File: winterfell/test/testFileClaimController.js (hunk 6 above)
Comment (qwwqwwq) [inline winterfell/test/testFileClaimController.js]: oh this is just to establish an authorized session for the other tests
